rpc/types_test.go

```diff
@@ -122,3 +123,33 @@ func TestBlockNumberOrHash_UnmarshalJSON(t *testing.T) {
 		}
 	}
 }
+
+func TestBlockNumberOrHash_WithNumber_MarshalAndUnmarshal(t *testing.T) {
+	tests := []struct {
+		name   string
+		number int64
+	}{
+		{"max", math.MaxInt64},
+		{"pending", int64(PendingBlockNumber)},
+		{"latest", int64(LatestBlockNumber)},
+		{"earliest", int64(EarliestBlockNumber)},
+	}
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			bnh := BlockNumberOrHashWithNumber(BlockNumber(test.number))
+			marshalled, err := json.Marshal(bnh)
+			if err != nil {
+				t.Fatal("cannot marshal:", err)
+			}
+			var unmarshalled BlockNumberOrHash
+			err = json.Unmarshal(marshalled, &unmarshalled)
+			if err != nil {
+				t.Fatal("cannot unmarshal:", err)
+			}
+			if !reflect.DeepEqual(bnh, unmarshalled) {
+				t.Fatalf("wrong result: expected %v, got %v", bnh, unmarshalled)
+			}
+		})
+	}
+}
```

Comment:
Might skip a layer of test nesting: 
```suggestion
	for _, test := range tests {
	    bnh := BlockNumberOrHashWithNumber(BlockNumber(test.number))
	    marshalled, err := json.Marshal(bnh)
	    if err != nil {
		    t.Fatal("cannot marshal:", err)
	    }
	    var unmarshalled BlockNumberOrHash
	    if err = json.Unmarshal(marshalled, &unmarshalled); err != nil {
		    t.Fatal("cannot unmarshal:", err)
	    }
	    if !reflect.DeepEqual(bnh, unmarshalled) {
		    t.Fatalf("wrong result: expected %v, got %v", bnh, unmarshalled)
	    }
	}
}
```

Reply:
test.name would become unused

Reply:
There are two pros of test nesting here:
1) failure related to one dataset doesn't suppress checking of other datasets,
2) in case of failure, the report clearly shows for which of data sets the code doesn't work correctly.

Reply:
```
--- FAIL: TestBlockNumberOrHash_WithNumber_MarshalAndUnmarshal (0.00s)
    --- FAIL: TestBlockNumberOrHash_WithNumber_MarshalAndUnmarshal/pending (0.00s)
        types_test.go:148: cannot unmarshal: json: cannot unmarshal object into Go value of type string
    --- FAIL: TestBlockNumberOrHash_WithNumber_MarshalAndUnmarshal/latest (0.00s)
        types_test.go:148: cannot unmarshal: json: cannot unmarshal object into Go value of type string
```